### PR TITLE
Fix single-quotes incorrectly escaped when dragging into script editor

### LINKED ISSUE #56263: `'` in file name gets incorrectly escaped when dragged to script editor
### Godot version

28174d5

### System information

W10

### Issue description

LwTQL1vpKh
I think it's supposed to escape in case of single quotes, but it's invalid when using doubles.

### Steps to reproduce

0. Use double quotes as default style
1. Have a file with `'` in name
2. Drag it to script editor

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by anvilfolk**
Fixes #56263. Fundamentally the same as godotengine/godot#57085.

Opted to leave ustring.cpp alone, as suggested by rrohak, and instead add a method to ScriptTextEditor that escapes a String for drag-n-drop into the script window according to single-quote preferences. It does that by de-escaping single-quotes that were escaped by String::c_escape, only when relevant.

### CODE REVIEW

**Comment by Chaosus** [comment]
We don't use sentences like "Fixes [something]" in commits and PR names, please rename it to reflect the change itself like: "Fix ' in file name gets incorrectly escaped when dragged to script editor" or shorter, don't forget to amend the commit name as well as the PR name.

**Comment by anvilfolk** [comment]
> We don't use sentences like `Fixes [link]` in commits and PR names, please rename it to reflect the change itself like: `Fix ' in file name gets incorrectly escaped when dragged to script editor` or shorter, don't forget to amend the commit name as well as the PR name.

Done, and apologies. A lot to keep track of and messed that up. Hopefully fixed!

File: editor/plugins/script_text_editor.cpp [hunk 1]
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
+		escaped = escaped.replace("\\'", "\'"); 
+	}
+
+	return escaped;
+}
+
 void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from) {
-	const String quote_style = EDITOR_GET("text_editor/completion/use_single_quotes") ? "'" : "\"";
+	const bool using_single_quotes = EDITOR_GET("text_editor/completion/use_single_quotes");
+	const String quote_style = using_single_quotes ? "'" : "\"";
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1621]: Can we remove the need for this and put `.quote` inside  `_escape_drop_data`?

File: editor/plugins/script_text_editor.cpp [hunk 2]
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
 	return nullptr;
 }
 
+
+String ScriptTextEditor::_escape_drop_data(const String &str) {
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1602]: Could we make this static?

File: editor/plugins/script_text_editor.cpp [hunk 3]
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
+	// This function escapes resource paths, node paths, or object properties
+	// according to String delimiter preferences.
+	
+	String escaped = str.c_escape();
+
+	// Resource paths, node paths and object properties cannot contain double-quotes
+	// so we do not bother reverting back the .replace from c_escape.
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1609]: This comment seems a bit out of place.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Looks good, couple small changes commented inline.

Applied the patch locally and it complained about a formatting issue would recommend setting up clang-format, can use the guide here: https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html

File: editor/plugins/script_text_editor.cpp (hunk 1 above)
Comment (anvilfolk) [reply editor/plugins/script_text_editor.cpp]: Yes, excellent suggestion! Changed _escape_drop_data to _quote_drop_data which already escapes and quotes as needed!

File: editor/plugins/script_text_editor.cpp (hunk 2 above)
Comment (anvilfolk) [reply editor/plugins/script_text_editor.cpp]: Done! No reason it shouldn't be indeed :)

File: editor/plugins/script_text_editor.cpp (hunk 3 above)
Comment (anvilfolk) [reply editor/plugins/script_text_editor.cpp]: Agreed, leftover from a previous understanding of the situation in which I thought filenames could have double-quotes. They cannot, so now we just don't bother worrying about them :)

**Comment by anvilfolk** [comment]
> Looks good, couple small changes commented inline.
> 
> Applied the patch locally and it complained about a formatting issue would recommend setting up clang-format, can use the guide here: https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html

Agreed and implemented all the changes, squashed commit and force-pushed. First time doing all this, but it looks like it might've worked as expected :)

I am using Visual Studio 2022 and it's supposed to apply the .clang-format file automatically (even warns me of it), but I went ahead and force-formatted the .h and .cpp file, so hopefully we won't have any issues!

File: editor/plugins/script_text_editor.h [hunk 4]
```diff
@@ -196,6 +196,7 @@ class ScriptTextEditor : public ScriptEditorBase {
 	void drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from);
 
 	String _get_absolute_path(const String &rel_path);
+	static String _quote_drop_data(const String &str);
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.h:199]: No need to put it in the header if it's static.

File: editor/plugins/script_text_editor.cpp [hunk 5]
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
+
+	// Unescape single quotes if user prefers double-quote delimited strings.
+	if (!using_single_quotes) {
+		escaped = escaped.replace("\\'", "\'"); 
+	}
+
+	// Add preferred quotes.
+	const String quote_style = using_single_quotes ? "'" : "\"";
+
+	return escaped.quote(quote_style);
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1619]:
```suggestion
	return escaped.quote(using_single_quotes ? "'" : "\"");
```

File: editor/plugins/script_text_editor.cpp [hunk 6]
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
-void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from) {
-	const String quote_style = EDITOR_GET("text_editor/completion/use_single_quotes") ? "'" : "\"";
 
+String ScriptTextEditor::_quote_drop_data(const String &str) {
+	// This function prepares a string for being "dropped" into the script editor.
+	// The string can be a resource path, node path or property name.
+
+	const bool using_single_quotes = EDITOR_GET("text_editor/completion/use_single_quotes");
+
+	// Escape any special characters in the string
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1608]: Useless comment as it duplicates the line of code.

File: editor/plugins/script_text_editor.cpp [hunk 7]
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
+	String escaped = str.c_escape();
+
+	// Unescape single quotes if user prefers double-quote delimited strings.
+	if (!using_single_quotes) {
+		escaped = escaped.replace("\\'", "\'"); 
+	}
+
+	// Add preferred quotes.
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1616]: Ditto.

File: editor/plugins/script_text_editor.cpp [hunk 8]
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
...
+String ScriptTextEditor::_quote_drop_data(const String &str) {
+	// This function prepares a string for being "dropped" into the script editor.
+	// The string can be a resource path, node path or property name.
+
+	const bool using_single_quotes = EDITOR_GET("text_editor/completion/use_single_quotes");
+
+	// Escape any special characters in the string
+	String escaped = str.c_escape();
+
+	// Unescape single quotes if user prefers double-quote delimited strings.
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1611]: Should explain the why a bit cleaner, something like:
```suggestion
	// If string is double quoted, there is no need to escape single quotes, so we can remove the extra escaping added above here.
```

File: editor/plugins/script_text_editor.h (hunk 4 above)
Comment (anvilfolk) [reply editor/plugins/script_text_editor.h:199]: I'm trying to figure out what the best way to define this type of method:
- It's a helper method for ScriptTextEditor
- It's not used outside of ScriptTextEditor (therefore shouldn't be in the .h file like you said)
- Someone in dev chat was saying it's uncommon to have methods that aren't associated with a class.

`String _quote_drop_data(const String &str) { ... }` seems to counter that last point.
`static String ScriptTextEditor::_quote_drop_data(const String &str) { ... }` doesn't compile without it also in the .h file
`static String _quote_drop_data(const String &str) { ... }` compiles but not entirely sure what it does, hehe

Haven't done C++ in almost 2 decades, so probably missing something here :)

Comment (anvilfolk) [reply editor/plugins/script_text_editor.h]: Pushed a change using the first method, which I think was your suggestion. Happy to change if it wasn't :)

Comment (Paulb23) [reply editor/plugins/script_text_editor.h]: Should be the latter one `static String _quote_drop_data(const String &str) { ... }` :)

It's from C to make the function only visible in this file.

It's not the most common pattern, but used a few times in the `ScriptEditors` for small utility methods like this, see `_find_script_node` above this one. There's a few more further up too.

Comment (anvilfolk) [reply editor/plugins/script_text_editor.h]: Changes submitted!

I was assuming C++, will look into static in C, thank you!

**Review by Paulb23 — APPROVED** [review]
LGTM

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Thank you for fixing this! 🎉

**Pull request merged**